go/extractor/project/project.go

```diff
@@ -42,109 +124,366 @@ func checkDirsNested(inputDirs []string) (string, bool) {
+func LoadGoModules(goModFilePaths []string) []*GoModule {
+	results := make([]*GoModule, len(goModFilePaths))
+
+	for i, goModFilePath := range goModFilePaths {
+		results[i] = new(GoModule)
+		results[i].Path = goModFilePath
+
+		modFileSrc, err := os.ReadFile(goModFilePath)
+
+		if err != nil {
+			log.Printf("Unable to read %s: %s.\n", goModFilePath, err.Error())
+			continue
 		}
-		if emitDiagnostics {
-			if nested {
-				diagnostics.EmitMultipleGoModFoundNested(goModPaths)
-			} else {
-				diagnostics.EmitMultipleGoModFoundNotNested(goModPaths)
-			}
+
+		modFile, err := modfile.ParseLax(goModFilePath, modFileSrc, nil)
+
+		if err != nil {
+			log.Printf("Unable to parse %s: %s.\n", goModFilePath, err.Error())
+			continue
 		}
-		return
+
+		results[i].Module = modFile
 	}
-	if emitDiagnostics {
-		if goModDirs[0] == "." {
-			diagnostics.EmitSingleRootGoModFound(goModPaths[0])
+
+	return results
+}
+
+// Given a path to a `go.work` file, this function attempts to parse the `go.work` file. If unsuccessful,
+// we attempt to discover `go.mod` files within subdirectories of the directory containing the `go.work`
+// file ourselves.
+func discoverWorkspace(workFilePath string) GoWorkspace {
+	log.Printf("Loading %s...\n", workFilePath)
+	baseDir := filepath.Dir(workFilePath)
+	workFileSrc, err := os.ReadFile(workFilePath)
+
+	if err != nil {
+		// We couldn't read the `go.work` file for some reason; let's try to find `go.mod` files ourselves
+		log.Printf("Unable to read %s, falling back to finding `go.mod` files manually:\n%s\n", workFilePath, err.Error())
+		return GoWorkspace{
+			BaseDir: baseDir,
+			Modules: LoadGoModules(findGoModFiles(baseDir)),
+			DepMode: GoGetWithModules,
+			ModMode: getModMode(GoGetWithModules, baseDir),
+		}
+	}
+
+	workFile, err := modfile.ParseWork(workFilePath, workFileSrc, nil)
+
+	if err != nil {
+		// The `go.work` file couldn't be parsed for some reason; let's try to find `go.mod` files ourselves
+		log.Printf("Unable to parse %s, falling back to finding `go.mod` files manually:\n%s\n", workFilePath, err.Error())
+		return GoWorkspace{
+			BaseDir: baseDir,
+			Modules: LoadGoModules(findGoModFiles(baseDir)),
+			DepMode: GoGetWithModules,
+			ModMode: getModMode(GoGetWithModules, baseDir),
+		}
+	}
+
+	// Get the paths of all of the `go.mod` files that we read from the `go.work` file.
+	goModFilePaths := make([]string, len(workFile.Use))
+
+	for i, use := range workFile.Use {
+		if filepath.IsAbs(use.Path) {
+			// TODO: This case might be problematic for some other logic (e.g. stray file detection)
+			goModFilePaths[i] = filepath.Join(use.Path, "go.mod")
 		} else {
-			diagnostics.EmitSingleNonRootGoModFound(goModPaths[0])
+			goModFilePaths[i] = filepath.Join(filepath.Dir(workFilePath), use.Path, "go.mod")
 		}
 	}
-	baseDir = goModDirs[0]
-	useGoMod = true
-	return
+
+	log.Printf("%s uses the following Go modules:\n%s\n", workFilePath, strings.Join(goModFilePaths, "\n"))
```

Comment:
Nice to have: It would be nice if there was a log message like this one, listing all the go modules, in the two cases above (can't read go.work and can't parse go.work).